origin.spec OKD RPM build readiness

LorbusChris opened the pull request:
- Prepared the specfile for builds of OKD packages
- Removed mentions of Tito

DanyC97 commented on origin.spec:
```diff
@@ -46,16 +46,16 @@
 %else
-%global package_name openshift
-%global product_name OpenShift
+%global package_name okd
```
aren't we breaking OCP here?

LorbusChris replied:
hm, on epel it'll still be openshift. See the lines above

LorbusChris replied:
This did actually break things in OCP. Should be fixed by the `.openshift` dist tag now

smarterclayton replied:
Let's just use package_name openshift for everything.  Do we *need* to call it OKD?

smarterclayton replied:
Oh, I see.  Yeah, let's call the variable `ocp_build`, keep the package name the same (no reason not to), and change product name based on ocp_build

DanyC97 commented on origin.spec:
```diff
@@ -7,15 +7,15 @@
 # os_git_vars needed to run hack scripts during rpm builds
 %{!?os_git_vars:
-%global os_git_vars OS_GIT_VERSION='' OS_GIT_COMMIT='' OS_GIT_MAJOR='' OS_GIT_MINOR='' OS_GIT_TREE_STATE=''
+%global os_git_vars OS_GIT_COMMIT='e795244084' OS_GIT_TREE_STATE='dirty' OS_GIT_VERSION='v4.1.0+e795244-1411-dirty' OS_GIT_MAJOR='4' OS_GIT_MINOR='1+' OS_GIT_PATCH='0' KUBE_GIT_MAJOR='1' KUBE_GIT_MINOR='14+' KUBE_GIT_COMMIT='e8fcd6a' KUBE_GIT_VERSION='v1.14.0+e795244084' ETCD_GIT_VERSION='v3.3.10' ETCD_GIT_COMMIT='27fc7e2' 
```
any reason why we added the version in too ? 
i may miss something here but if i remember in the past the hack/* script was dynamically working out the version ... not sure if this breaks OCP

LorbusChris replied:
Afaict this is just what OCP passes as os_git_vars right now and line just hasn't been kept up to date before. Putting them all in one var does obscure things here a bit.

DanyC97 replied:
_line just hasn't been kept up to date before_ => that is what i was afraid you will say after i posted my comment, thanks for the info !

DanyC97 commented on hack/build-rpms.sh:
```diff
@@ -22,7 +21,7 @@ os::build::setup_env
 
 if [[ "${OS_ONLY_BUILD_PLATFORMS:-}" == 'linux/amd64' ]]; then
 	# when the user is asking for only Linux binaries, we will
-	# furthermore not build cross-platform clients in tito
```
am i right in thinking that currently tito is no longer in used (e.g it was in 3.x but world moved on since then? )

LorbusChris replied:
I will need some clarification on where it is still used today, if at all.

smarterclayton replied:
I’m not aware of any use in this repo, and it hasn’t been updated in a year or more.

LorbusChris commented:
Verification is slightly tricky, because one revision of the specfile can only contain a git ref to the previous commit, before the spec itself was commited.
Changed the script to verify the spec against HEAD~1 (to be done after the commit) and removed the it from the `make verify` target for now. This isn't ideal.

To me, the best flow would look something like this in order to make sure the spec is updated with a git ref that contains all other updates:
1 make update (as is now, !spec),
2 git commit,
3 make update spec,
4 git commit
5 make verify all

I think Packit doesnt require the right commit and version to be set (at least on PRs). Will investigate.

In the meantime, does anybody have an idea how to solve this elegantly?

smarterclayton commented on origin.spec:
```diff
@@ -102,9 +103,14 @@ Summary:        %{product_name} Kubernetes server commands
+%if 0%{?epel}
+Conflicts:      okd-hyperkube
+%else
+Conflicts:      openshift-hyperkube
```
Why add this?

LorbusChris replied:
Intention was to signal that the user can't install both at the same time due to conflicting files.

However, on second thought, as the openshift-hypershift rpm is not distributed stand-alone, I'd be fine with removing it again altogether because they might not ever ever come across each other.
Should I remove?

DanyC97 replied:
@LorbusChris @smarterclayton beat me on this ;) after our chat over the weekend i started to dig a bit and found https://github.com/openshift/release/pull/4466 & https://github.com/openshift/installer/pull/2131 hence the hint i guess.

LorbusChris replied:
Above links have something to do with removing `hypershift`.  This is about `hyperkube`, though. I don't have a strong opinion about this one and can remove it, but my interpretation of https://docs.fedoraproject.org/en-US/packaging-guidelines/Conflicts/#_implicit_conflicts would still be to add it.

DanyC97 replied:
sorry, ignore what i've said, monday morning w/o coffee.

sgreene570 replied:
> This is about `hyperkube`, though. I don't have a strong opinion about this one and can remove it, but my interpretation of https://docs.fedoraproject.org/en-US/packaging-guidelines/Conflicts/#_implicit_conflicts would still be to add it.

+ 1

Is there any disadvantage to being verbose and including the conflict in this situation?

smarterclayton replied:
In this case we shouldn't have `atomic-openshift` anymore, OKD can use `openshift` for package names.

smarterclayton replied:
`openshift-hyperkube`

smarterclayton commented on origin.spec:
```diff
@@ -142,16 +148,10 @@ OS_ONLY_BUILD_PLATFORMS="${BUILD_PLATFORM}" %{os_git_vars} OS_BUILD_RELEASE_ARCH
-  echo "+++ INSTALLING ${bin}"
-  install -p -m 755 _output/local/bin/${PLATFORM}/${bin} %{buildroot}%{_bindir}/${bin}
-done
+echo "+++ INSTALLING hyperkube"
```
Leave the loop, we may have additional binaries in this package.

smarterclayton commented on origin.spec:
```diff
@@ -46,16 +46,16 @@
-%{!?version: %global version 0.0.1}
-%{!?release: %global release 1}
+%{!?version: %global version 4.1.0}
+%{!?release: %global release 1.1417.98b35a9}
```
Are you generating this?  It would change on ever commit if so which won’t work.

LorbusChris replied:
Yeah..I realized after the fact that committing the vars can never work with CI.
The right way is probably to have a proper make target for OKD builds and pass the vars at build time in a proper CI pipeline.

LorbusChris replied:
Removed the spec update generation.

smarterclayton commented on origin.spec:
```diff
@@ -46,16 +46,16 @@
 
-%{!?version: %global version 0.0.1}
-%{!?release: %global release 1}
+%{!?version: %global version 4.1.0}
```
I don’t think we should be tying this version to a top level version, the current plan is that component version numbers are completely component dependent and likely not to be managed under the covers (the rpm may vary wildly over time).  Probably better to start at 4.0.0

LorbusChris replied:
This was also taken from the auto-generation. Set to 4.0.0 now.

smarterclayton commented:
Looks like this broke the ostree override for PR testing

LorbusChris commented:
/retest

LorbusChris commented:
I just noticed that we shouldn't inject `Provides: bundled(x)`, but rather `BuildRequires: bundled(x)` as per https://docs.fedoraproject.org/en-US/packaging-guidelines/Golang/#_buildrequires, and there is a macro for that:
```
%generate_buildrequires
%go_generate_buildrequires
```
Will try to get that to work next.

LorbusChris commented on hack/build-rpms.sh:
```diff
@@ -58,12 +58,14 @@ if [[ -n "${dirty}" && "${OS_GIT_TREE_STATE}" == "dirty" ]]; then
 		--define "skip_prep 1" \
-		--define "skip_dist 1" \
+		--define "skip_dist ${SKIP_DIST:-1}" \
+		--define "ose_build ${OSE_BUILD:-1}" \
```
ose or ocp build?

smarterclayton replied:
`ocp_build` if you have too.

LorbusChris replied:
won't be necessary if we can use `openshift` name schema

LorbusChris commented on hack/build-rpms.sh:
```diff
@@ -72,16 +74,19 @@ else
+		--define "release ${OS_RPM_RELEASE}" \
 		--define "commit ${OS_GIT_COMMIT}" \
 		--define "os_git_vars ${OS_RPM_GIT_VARS}" \
-		--define 'dist .el7' --define "_topdir ${rpm_tmp_dir}"
```
can I safely remove the `dist` override here? Looks like it it is skipped anyway by `skip_dist`

smarterclayton commented on origin.spec:
```diff
@@ -46,15 +46,15 @@
-%global package_name openshift
-%global product_name OpenShift
+%global package_name okd
+%global product_name OKD
```
What are the rules on Product Name in general?

LorbusChris replied:
not sure. If we could reuse the `openshift` naming scheme across the board, we wouldn't need the ose/ocp distinction in rpmbuild

smarterclayton replied:
Yeah, let's reuse the openshift naming scheme across the board for now.  We call component pieces OpenShift and they are allowed to use the word.  It's only the top level that is officially called OKD for Reasons(tm) but we allow the use of the term to refer to the code base.

LorbusChris commented:
/retest

LorbusChris commented:
/retest

LorbusChris commented:
/retest

smarterclayton commented:
/lgtm

If this has an impact on ART builds we'll revert
